### PR TITLE
[WIP] cnf can be too expensive so we need a backup

### PR DESCRIPTION
**Pull request by tobymao**
compute normalization distance as a heuristic for how expensive it will
be. thus, not all where statements will be in CNF, we need another
strategy.

additionally move all predicate mutations to pushdown predicates and out
of decorrelation and expansion.

### CODE REVIEW

File: sqlglot/expressions.py [hunk 1]
```diff
@@ -580,6 +578,46 @@ def is_int(self):
...
 
+    @property
+    def kind(self):
+        return self.text("kind").upper()
+
+    def on(self, *expressions, append=True, dialect=None, parser_opts=None, copy=True):
+        """
+        Append to or set the ON expressions.
+
+        Example:
```
Comment (barakalon) [inline sqlglot/expressions.py:589]: Example copy and pasted from somewhere - needs to be updated

File: sqlglot/expressions.py [hunk 2]
```diff
@@ -580,6 +578,46 @@ def is_int(self):
...
+            *expressions,
+            instance=self,
+            arg="on",
+            append=append,
+            dialect=dialect,
+            parser_opts=parser_opts,
+            copy=copy,
+        )
+
+        if join.kind == "CROSS":
```
Comment (barakalon) [inline sqlglot/expressions.py:616]: What's this about? Should this be `if join.kind == "CROSS" and not expressions`?

File: sqlglot/optimizer/normalize.py [hunk 3]
```diff
@@ -1,9 +1,9 @@
 from sqlglot.helper import while_changing
-from sqlglot.optimizer.simplify import simplify
+from sqlglot.optimizer.simplify import simplify, compare_and_prune, flatten
 import sqlglot.expressions as exp
 
 
-def normalize(expression, dnf=False):
+def normalize(expression, dnf=False, max_distance=128):
```
Comment (barakalon) [inline sqlglot/optimizer/normalize.py:6]: `max_distance` deserves a spot in the docstring

File: sqlglot/optimizer/normalize.py [hunk 4]
```diff
@@ -20,11 +20,65 @@ def normalize(expression, dnf=False):
...
+        int: difference
+    """
+    return sum(_predicate_lengths(expression, dnf)) - (
+        len(list(expression.find_all(exp.Connector))) + 1
+    )
+
+
+def _predicate_lengths(expression, dnf):
+    """
+    Returns a list of predicates lengths when expanded to normalized form.
```
Comment (barakalon) [inline sqlglot/optimizer/normalize.py:63]: nit: "predicate lengths"

File: sqlglot/expressions.py (hunk 2 above)
Comment (tobymao) [reply sqlglot/expressions.py:616]: if the current join is a cross join, and you call this method it is no longer a cross join

File: sqlglot/optimizer/pushdown_predicates.py [hunk 5]
```diff
@@ -18,58 +20,116 @@ def pushdown_predicates(expression):
...
 
-            if len(sources) != 1:
-                continue
+        if cnf_like:
+            for predicate in predicates:
+                for node in nodes_for_predicate(predicate, scope).values():
+                    if isinstance(node, exp.Join):
+                        predicate.replace(exp.TRUE)
+                        node.on(predicate, copy=False)
+                        break
```
Comment (barakalon) [inline sqlglot/optimizer/pushdown_predicates.py:47]: Why break?

File: sqlglot/optimizer/pushdown_predicates.py [hunk 6]
```diff
@@ -18,58 +20,116 @@ def pushdown_predicates(expression):
...
+                    if isinstance(node, exp.Select):
+                        predicate.replace(exp.TRUE)
+                        node.where(replace_aliases(node, predicate), copy=False)
+        else:
+            pushdown = set()
```
Comment (barakalon) [inline sqlglot/optimizer/pushdown_predicates.py:52]: This block is very hard to follow. 

I know complexity is complex... but maybe we can throw some comments here and there?

File: sqlglot/optimizer/scope.py [hunk 7]
```diff
@@ -154,18 +154,22 @@ def selected_sources(self):
...
-            referenced_names.append(derived_table.alias)
+            referenced_names.append((derived_table.alias, derived_table.this))
 
         result = {}
 
-        for name in referenced_names:
+        for name, node in referenced_names:
             if name in self.sources:
-                result[name] = self.sources[name]
+                result[name] = (node, self.sources[name])
```
Comment (barakalon) [inline sqlglot/optimizer/scope.py:172]: We need to update this methods docstring

File: tests/fixtures/optimizer/pushdown_predicates.sql [hunk 8]
```diff
@@ -1,17 +1,20 @@
...
-
-SELECT x.a FROM (SELECT * FROM x) AS x JOIN y WHERE x.a = 1 AND x.b = 1 OR y.a = 1;
-SELECT x.a FROM (SELECT * FROM x) AS x JOIN y WHERE x.a = 1 AND x.b = 1 OR y.a = 1;
-
-SELECT x.a FROM (SELECT x.a AS a, x * 1 AS c FROM x) AS x WHERE x.c = 1;
-SELECT x.a FROM (SELECT x.a AS a, x * 1 AS c FROM x WHERE x * 1 = 1) AS x WHERE TRUE;
-
-SELECT x.a FROM (SELECT x.a AS a, x * 1 AS c FROM x) AS x WHERE x.c = 1 or x.c = 2;
-SELECT x.a FROM (SELECT x.a AS a, x * 1 AS c FROM x WHERE x * 1 = 1 OR x * 1 = 2) AS x WHERE TRUE;
+--SELECT x.a AS a FROM (SELECT x.a FROM x AS x) AS x JOIN y WHERE x.a = 1 AND x.b = 1 AND y.a = 1;
```
Comment (barakalon) [inline tests/fixtures/optimizer/pushdown_predicates.sql:1]: Commented on purpose?

File: tests/fixtures/optimizer/tpc-h/tpc-h.sql [hunk 9]
```diff
@@ -978,39 +977,49 @@ LIMIT 100;
 --------------------------------------
 -- TPC-H 19
 --------------------------------------
---select
```
Comment (barakalon) [inline tests/fixtures/optimizer/tpc-h/tpc-h.sql:981]: Commented on purpose?

**Review by barakalon — APPROVED** [review]
Oomph, this is nuts.

File: sqlglot/expressions.py (hunk 2 above)
Comment (barakalon) [reply sqlglot/expressions.py:616]: Ah yeah, doy.

File: sqlglot/optimizer/pushdown_predicates.py (hunk 5 above)
Comment (tobymao) [reply sqlglot/optimizer/pushdown_predicates.py:47]: we only want to push it down to one join

Comment (barakalon) [reply sqlglot/optimizer/pushdown_predicates.py:47]: I know some folks are anti-comments...
But I think little comments like this are helpful, especially with gnarly code like this.

File: tests/fixtures/optimizer/pushdown_predicates.sql (hunk 8 above)
Comment (tobymao) [reply tests/fixtures/optimizer/pushdown_predicates.sql:1]: nope
